src/core/thread/data_poll_manager.cpp

```diff
@@ -392,20 +392,29 @@ otError DataPollManager::StopFastPolls(void)
 
 void DataPollManager::ScheduleNextPoll(PollPeriodSelector aPollPeriodSelector)
 {
+    uint32_t now, interval;
+
     if (aPollPeriodSelector == kRecalculatePollPeriod)
     {
         mPollPeriod = CalculatePollPeriod();
     }
 
-    if (mTimer.IsRunning())
+    now      = TimerMilli::GetNow();
+    interval = mPollPeriod;
+
+    if (!mTimer.IsRunning())
     {
-        mTimer.StartAt(mTimerStartTime, mPollPeriod);
+        mTimerStartTime = now;
     }
-    else
+    // Ensure at least kMinPollPeriod (`OPENTHREAD_CONFIG_MINIMUM_POLL_PERIOD`, 10ms by default)
+    // before sending data poll.
+    else if (TimerScheduler::IsStrictlyBefore((mTimerStartTime + mPollPeriod - kMinPollPeriod), now))
```

Comment:
I think maybe safer if we trigger this check only if the poll interval does change?

Existing implementation let us call `ScheduleNextPoll()` again (when there is no interval change) and it would have no impact (it would restart timer from same start time with same period so no change). Not sure if this behavior is used (relied on) but safer to keep it.